File: java/src/main/java/com/meteroid/Webhook.java
package com.meteroid;

import com.standardwebhooks.exceptions.WebhookVerificationException;

import java.net.http.HttpHeaders;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Utility class for verifying Meteroid webhook signatures.
 *
 * <p>This class wraps the standard-webhooks library to provide webhook signature
 * verification for Meteroid webhooks. It supports both Svix-branded headers (svix-*)
 * and standard webhook headers (webhook-*).
 *
 * <p>Header lookup order:
 * <ol>
 *   <li>svix-id, svix-signature, svix-timestamp (Svix branded)</li>
 *   <li>webhook-id, webhook-signature, webhook-timestamp (standard)</li>
 * </ol>
 *
 * <p>Example usage:
 * <pre>{@code
 * Webhook webhook = new Webhook("whsec_your_secret_here");
 *
 * // Verify webhook from HTTP headers
 * webhook.verify(payload, headers);
 * }</pre>
 *
 * @see <a href="https://github.com/standard-webhooks/standard-webhooks">Standard Webhooks</a>
 */
public final class Webhook {
    public static final String SECRET_PREFIX = "whsec_";

    // Svix-branded header keys (primary)
    public static final String SVIX_MSG_ID_KEY = "svix-id";
    public static final String SVIX_MSG_SIGNATURE_KEY = "svix-signature";
    public static final String SVIX_MSG_TIMESTAMP_KEY = "svix-timestamp";

    // Standard webhook header keys (fallback)
    public static final String WEBHOOK_MSG_ID_KEY = "webhook-id";
    public static final String WEBHOOK_MSG_SIGNATURE_KEY = "webhook-signature";
    public static final String WEBHOOK_MSG_TIMESTAMP_KEY = "webhook-timestamp";

    private final com.standardwebhooks.Webhook delegate;

    /**
     * Create a new Webhook verifier with the given secret.
     *
     * @param secret The webhook signing secret. May include the "whsec_" prefix.
     */
    public Webhook(final String secret) {
        this.delegate = new com.standardwebhooks.Webhook(secret);
    }

    /**
     * Create a new Webhook verifier with the given secret as raw bytes.
     *
     * @param secret The webhook signing secret as raw bytes (Base64-decoded).
     */
    public Webhook(final byte[] secret) {
        this.delegate = new com.standardwebhooks.Webhook(secret);
    }

    /**
     * Verify a webhook payload using HttpHeaders.
     *
     * <p>This method validates:
     * <ul>
     *   <li>The timestamp is within the allowed tolerance (5 minutes)</li>
     *   <li>The signature matches the expected HMAC-SHA256 signature</li>
     * </ul>
     *
     * <p>Accepts both svix-* and webhook-* header formats.
     *
     * @param payload The raw webhook payload body as a string
     * @param headers The HTTP headers from the webhook request
     * @throws WebhookVerificationException if verification fails
     */
    public void verify(final String payload, final HttpHeaders headers)
            throws WebhookVerificationException {
        Map<String, List<String>> headerMap = normalizeHeaders(headers.map());
        delegate.verify(payload, headerMap);
    }

    /**
     * Verify a webhook payload using a header map.
     *
     * <p>This method is useful when you have headers as a Map rather than HttpHeaders.
     * Accepts both svix-* and webhook-* header formats.
     *
     * @param payload The raw webhook payload body as a string
     * @param headers Map of header names to lists of header values
     * @throws WebhookVerificationException if verification fails
     */
    public void verify(final String payload, final Map<String, List<String>> headers)
            throws WebhookVerificationException {
        Map<String, List<String>> normalizedHeaders = normalizeHeaders(headers);
        delegate.verify(payload, normalizedHeaders);
    }

    /**
     * Sign a webhook payload (useful for testing).
     *
     * @param msgId The message ID
     * @param timestamp Unix timestamp in seconds
     * @param payload The payload to sign
     * @return The signature in "v1,{base64-signature}" format
     * @throws com.standardwebhooks.exceptions.WebhookSigningException if signing fails
     */
    public String sign(final String msgId, final long timestamp, final String payload)
            throws com.standardwebhooks.exceptions.WebhookSigningException {
        return delegate.sign(msgId, timestamp, payload);
    }

    /**
     * Normalize headers by converting svix-* headers to webhook-* format.
     * If svix-* headers are present, they take precedence over webhook-* headers.
     */
    private Map<String, List<String>> normalizeHeaders(final Map<String, List<String>> headers) {
        Map<String, List<String>> normalized = new HashMap<>();

        // Copy all headers with lowercase keys
        for (Map.Entry<String, List<String>> entry : headers.entrySet()) {
            normalized.put(entry.getKey().toLowerCase(), new ArrayList<>(entry.getValue()));
        }

        // Map svix-* headers to webhook-* headers (svix takes precedence)
        mapHeader(normalized, SVIX_MSG_ID_KEY, WEBHOOK_MSG_ID_KEY);
        mapHeader(normalized, SVIX_MSG_SIGNATURE_KEY, WEBHOOK_MSG_SIGNATURE_KEY);
        mapHeader(normalized, SVIX_MSG_TIMESTAMP_KEY, WEBHOOK_MSG_TIMESTAMP_KEY);

        return normalized;
    }

    /**
     * If the source header exists, copy its value to the target header.
     */
    private void mapHeader(Map<String, List<String>> headers, String source, String target) {
        List<String> sourceValue = headers.get(source);
        if (sourceValue != null && !sourceValue.isEmpty()) {
            headers.put(target, sourceValue);
        }
    }
}

File: typescript/src/webhook.ts
import { Webhook as StdWh, WebhookVerificationError } from "standardwebhooks";

// Re-export for users to catch specific verification errors
export { WebhookVerificationError };

export interface WebhookOptions {
  format?: "raw";
}

export interface WebhookRequiredHeaders {
  "webhook-id": string;
  "webhook-signature": string;
  "webhook-timestamp": string;
}

const HEADER_ID = "webhook-id";
const HEADER_SIGNATURE = "webhook-signature";
const HEADER_TIMESTAMP = "webhook-timestamp";

const SVIX_ID = "svix-id";
const SVIX_SIGNATURE = "svix-signature";
const SVIX_TIMESTAMP = "svix-timestamp";

/**
 * Webhook verification utility for verifying incoming webhook payloads.
 *
 * Uses the standardwebhooks library for HMAC-SHA256 signature verification.
 *
 * Two header naming conventions are accepted, matching the Rust, Java and Python
 * SDKs:
 *
 * - **Standard Webhooks** (preferred): `webhook-id`, `webhook-signature`,
 *   `webhook-timestamp`
 * - **Svix** (for compatibility): `svix-id`, `svix-signature`, `svix-timestamp`
 *
 * A `webhook-*` header always takes precedence over its `svix-*` counterpart when
 * both are present.
 */
export class Webhook {
  private readonly inner: StdWh;

  /**
   * Create a new Webhook verifier.
   *
   * @param secret The webhook signing secret (base64 encoded or prefixed with "whsec_")
   * @param options Optional configuration
   */
  constructor(secret: string | Uint8Array, options?: WebhookOptions) {
    this.inner = new StdWh(secret, options);
  }

  /**
   * Verify a webhook payload and return the parsed JSON body.
   *
   * @param payload The raw webhook payload (string or Buffer)
   * @param headers The webhook headers (must include `webhook-id`,
   *   `webhook-signature` and `webhook-timestamp`, or their `svix-*` aliases)
   * @returns The verified and parsed webhook payload
   * @throws WebhookVerificationError if verification fails
   */
  public verify(
    payload: string | Buffer,
    headers_: WebhookRequiredHeaders | Record<string, string>
  ): unknown {
    // Normalize headers to lowercase; header names are case-insensitive but a
    // plain object is not.
    const headers: Record<string, string> = {};
    for (const key of Object.keys(headers_)) {
      headers[key.toLowerCase()] = (headers_ as Record<string, string>)[key];
    }

    // Copy each `svix-*` header onto its `webhook-*` counterpart, but only when
    // the `webhook-*` one is absent: Standard Webhooks headers win.
    copyFallback(headers, HEADER_ID, SVIX_ID);
    copyFallback(headers, HEADER_SIGNATURE, SVIX_SIGNATURE);
    copyFallback(headers, HEADER_TIMESTAMP, SVIX_TIMESTAMP);

    return this.inner.verify(payload, headers);
  }

  /**
   * Sign a webhook payload.
   *
   * @param msgId Unique message identifier
   * @param timestamp Timestamp for the signature
   * @param payload The payload to sign (string or Buffer)
   * @returns The signature string
   */
  public sign(msgId: string, timestamp: Date, payload: string | Buffer): string {
    return this.inner.sign(msgId, timestamp, payload);
  }
}

function copyFallback(
  headers: Record<string, string>,
  primary: string,
  fallback: string
): void {
  if (headers[primary] === undefined && headers[fallback] !== undefined) {
    headers[primary] = headers[fallback];
  }
}
